File: server/engine/report_engine.js
const PDFDocument = require('pdfkit');
const fs = require('fs');
const path = require('path');
const sharp = require('sharp');

const translations = {
    en: {
        title: "Paddy Health & Disease Report",
        date: "Date",
        field: "Field",
        risk_score: "Risk Score",
        dominant_disease: "Dominant Disease",
        severity: "Severity",
        status: "Status",
        where_problem: "Where is the problem? (Infection Map)",
        what_to_do: "What to do now? (Recommendations)",
        action: "Action",
        treatment: "Treatment",
        dosage: "Dosage",
        spray_window: "Best Spray Time",
        why_suggestion: "Why this suggestion?",
        safety_note: "Safety Note",
        analyzed_view: "AI Vision: Analyzed Drone View",
        cost_breakdown: "Financial: Cost Estimation Breakdown",
        medicine: "Recommended Medicine",
        total_cost: "Total Application Cost (Est.)",
        levels: {
            CRITICAL: "CRITICAL",
            HIGH: "HIGH",
            MEDIUM: "MEDIUM",
            LOW: "LOW"
        }
    },
    te: {
        title: "వరి ఆరోగ్యం మరియు తెగులు నివేదిక",
        date: "తేదీ",
        field: "పొలం",
        risk_score: "ప్రమాద స్థాయి",
        dominant_disease: "ప్రధాన తెగులు",
        severity: "తీవ్రత",
        status: "స్థితి",
        where_problem: "సమస్య ఎక్కడ ఉంది? (మ్యాప్)",
        what_to_do: "ఇప్పుడు ఏమి చేయాలి? (సూచనలు)",
        action: "చేయవలసిన పని",
        treatment: "చికిత్స",
        dosage: "మోతాదు",
        spray_window: "పిచికారీ చేయడానికి ఉత్తమ సమయం",
        why_suggestion: "ఈ సూచన ఎందుకు?",
        safety_note: "భద్రతా గమనిక",
        analyzed_view: "AI దృష్టి: విశ్లేషించబడిన చిత్రం",
        cost_breakdown: "ఖర్చు అంచనా వివరాలు",
        medicine: "సిఫార్సు చేయబడిన మందు",
        total_cost: "మొత్తం ఖర్చు (అంచనా)",
        levels: {
            CRITICAL: "అత్యంత ప్రమాదకరం",
            HIGH: "ప్రమాదకరం",
            MEDIUM: "మధ్యస్థం",
            LOW: "తక్కువ"
        }
    }
};

class ReportEngine {
    constructor() {
        this.reportsDir = path.join(__dirname, '../uploads/reports');
        if (!fs.existsSync(this.reportsDir)) {
            fs.mkdirSync(this.reportsDir, { recursive: true });
        }
    }

    /**
     * Generate a PDF Report
     */
    async generatePDF(data, lang = 'en') {
        const t = translations[lang] || translations.en;
        const reportId = `report_${Date.now()}`;
        const filePath = path.join(this.reportsDir, `${reportId}.pdf`);

        return new Promise((resolve, reject) => {
            const doc = new PDFDocument({ margin: 50, size: 'A4' });
            const stream = fs.createWriteStream(filePath);

            doc.pipe(stream);

            // --- Header ---
            this.drawHeader(doc, t, data);

            // --- Risk Summary ---
            this.drawRiskSection(doc, t, data);

            // --- Analyzed Image (New) ---
            this.drawAnalyzedImageSection(doc, t, data);

            // --- Map Section (Simulated Grid) ---
            this.drawMapSection(doc, t, data);

            // --- Recommendation Section ---
            this.drawRecommendationSection(doc, t, data);

            // --- Cost Estimation (New) ---
            this.drawCostSection(doc, t, data);

            // --- Why & Safety ---
            this.drawFooter(doc, t, data);

            doc.end();

            stream.on('finish', () => resolve({ reportId, filePath }));
            stream.on('error', reject);
        });
    }

    drawHeader(doc, t, data) {
        doc.fillColor('#2d3436').fontSize(24).text(t.title, { align: 'center' });
        doc.moveDown(0.5);
        doc.fontSize(12).fillColor('#636e72').text(`${t.date}: ${new Date().toLocaleDateString()} | ${t.field}: ${data.field_name || 'My Farm'}`, { align: 'center' });
        doc.moveDown(1.5);
        doc.path('M 50 ' + doc.y + ' L 545 ' + doc.y).stroke('#dfe6e9');
        doc.moveDown(1.5);
    }

    drawRiskSection(doc, t, data) {
        const riskLevel = data.risk_level || 'LOW';
        const score = data.risk_score || 0;
        const colors = { CRITICAL: '#d63031', HIGH: '#e17055', MEDIUM: '#fdcb6e', LOW: '#00b894' };

        doc.fontSize(18).fillColor('#2d3436').text(`${t.risk_score}: `, { continued: true });
        doc.fillColor(colors[riskLevel]).text(`${score}/100 (${t.levels[riskLevel]})`);
        doc.moveDown(0.5);
        doc.fontSize(14).fillColor('#2d3436').text(`${t.dominant_disease}: ${data.dominant_disease || 'None Detected'}`);
        doc.moveDown(1.5);
    }

    /**
     * Fix paths that might be from a different folder name (e.g. 'agriculture new' vs 'agriculture new (1)')
     */
    fixImagePath(p) {
        if (!p) return null;
        if (fs.existsSync(p)) return p;

        // Try resolving relative to current root
        const filename = path.basename(p);
        const localPath = path.join(__dirname, '../uploads', filename);
        if (fs.existsSync(localPath)) return localPath;

        const uploadsPath = path.join(__dirname, '../uploads/reports', filename);
        if (fs.existsSync(uploadsPath)) return uploadsPath;

        return null; // Still not found
    }

    drawAnalyzedImageSection(doc, t, data) {
        const imagePath = this.fixImagePath(data.annotated_image);
        if (!imagePath) return;

        doc.fontSize(16).fillColor('#2d3436').text(t.analyzed_view);
        doc.moveDown(0.5);

        try {
            doc.image(imagePath, {
                fit: [500, 250],
                align: 'center'
            });
            doc.moveDown(10); // Provide enough space after image
        } catch (err) {
            console.error("PDF Image Error:", err);
        }

        doc.moveDown(1);
    }

    drawMapSection(doc, t, data) {
        doc.fontSize(16).fillColor('#2d3436').text(t.where_problem);
        doc.moveDown(1);

        const startX = 50;
        const startY = doc.y;
        const cellSize = 15;
        const rows = 10;
        const cols = 32;

        doc.rect(startX, startY, cols * cellSize, rows * cellSize).fill('#f1f2f6');
        const infectedTiles = data.infected_tiles || [[1, 2], [1, 3], [5, 5], [10, 8], [11, 8], [12, 1]];
        doc.fillColor('#ff7675');
        infectedTiles.forEach(tile => {
            doc.rect(startX + tile[0] * cellSize, startY + tile[1] * cellSize, cellSize, cellSize).fill();
        });

        doc.strokeColor('#ced4da').lineWidth(0.5);
        for (let i = 0; i <= cols; i++) doc.moveTo(startX + i * cellSize, startY).lineTo(startX + i * cellSize, startY + rows * cellSize).stroke();
        for (let i = 0; i <= rows; i++) doc.moveTo(startX, startY + i * cellSize).lineTo(startX + cols * cellSize, startY + i * cellSize).stroke();

        doc.y = startY + (rows * cellSize) + 30;
    }

    drawRecommendationSection(doc, t, data) {
        const rec = data.recommendation || {};
        doc.fontSize(16).fillColor('#2d3436').text(t.what_to_do);
        doc.moveDown(0.5);

        doc.fontSize(12).fillColor('#2d3436');
        doc.text(`• ${t.action}: ${rec.action || 'Monitoring'}`);
        doc.text(`• ${t.treatment}: ${rec.treatment?.product_name || 'Tricyclazole 75% WP'}`);
        doc.text(`• ${t.dosage}: ${rec.treatment?.dosage_per_acre || '120g in 200L water'}`);
        doc.text(`• ${t.spray_window}: ${rec.best_spray_window?.[0]?.time || '06:00 AM - 09:00 AM'}`);

        doc.moveDown(1.5);
    }

    drawCostSection(doc, t, data) {
        doc.fontSize(16).fillColor('#2d3436').text(t.cost_breakdown);
        doc.moveDown(0.5);

        // Fallback demo data if real cost info is missing
        const cost = data.cost_data || {
            medicine_name: "Tricyclazole (Beam/Amistar)",
            medicine_cost: 1550,
            equipment_cost: 350,
            application_cost: 450,
            total_min: 2350,
            total_max: 2800
        };

        const startX = 60;

        doc.fontSize(11).fillColor('#636e72');
        doc.text(`${t.medicine}:`, startX, doc.y);
        doc.fillColor('#2d3436').text(`${cost.medicine_name}`, startX + 180, doc.y - 13);
        doc.moveDown(0.3);

        doc.fillColor('#636e72').text("Estimated Medicine Cost:", startX, doc.y);
        doc.fillColor('#2d3436').text(`₹${cost.medicine_cost}`, startX + 180, doc.y - 13);
        doc.moveDown(0.3);

        doc.fillColor('#636e72').text("Labor & Machinery:", startX, doc.y);
        doc.fillColor('#2d3436').text(`₹${(cost.equipment_cost || 0) + (cost.application_cost || 0)}`, startX + 180, doc.y - 13);
        doc.moveDown(0.8);

        doc.fontSize(14).fillColor('#2d3436').text(`${t.total_cost}: `, { continued: true });
        doc.fillColor('#d63031').text(`₹${cost.total_min} - ₹${cost.total_max}`);

        doc.moveDown(1.5);
    }

    drawFooter(doc, t, data) {
        doc.fontSize(14).fillColor('#2d3436').text(t.why_suggestion);
        (data.explainability || []).slice(0, 3).forEach(reason => {
            doc.fontSize(11).fillColor('#636e72').text(`- ${reason}`);
        });

        doc.moveDown(1.5);
        doc.fontSize(12).fillColor('#d63031').text(`${t.safety_note}: ${data.recommendation?.constraints_analysis || "Avoid spraying in high wind (>15km/h) or upcoming rain."}`);
    }

    /**
     * Create WhatsApp-friendly image cards using Sharp
     */
    async generateWhatsAppCards(data, lang = 'en') {
        const t = translations[lang] || translations.en;
        const cardId = `card_${Date.now()}`;
        const portraitId = `portrait_${Date.now()}`;
        const squarePath = path.join(this.reportsDir, `${cardId}.png`);
        const portraitPath = path.join(this.reportsDir, `${portraitId}.png`);

        const score = data.risk_score || 0;
        const riskColor = score > 70 ? '#d63031' : (score > 40 ? '#e17055' : '#00b894');

        // --- Square Card (1080x1080) ---
        const squareSvg = `
        <svg width="1080" height="1080" xmlns="http://www.w3.org/2000/svg">
            <rect width="100%" height="100%" fill="white"/>
            <rect width="100%" height="200" fill="${riskColor}"/>
            <text x="50%" y="120" font-family="Arial" font-size="60" fill="white" text-anchor="middle" font-weight="bold">${t.title.replace(/&/g, '&amp;')}</text>
            <text x="50%" y="350" font-family="Arial" font-size="120" fill="${riskColor}" text-anchor="middle" font-weight="bold">${score}</text>
            <text x="50%" y="420" font-family="Arial" font-size="40" fill="#636e72" text-anchor="middle">${t.risk_score}</text>
            <text x="50%" y="1000" font-family="Arial" font-size="30" fill="#b2bec3" text-anchor="middle">Generated by PaddyPulse AI • ${new Date().toLocaleDateString()}</text>
        </svg>`;

        // --- Portrait Story (1080x1920) ---
        const portraitSvg = `
        <svg width="1080" height="1920" xmlns="http://www.w3.org/2000/svg">
            <rect width="100%" height="100%" fill="#f1f2f6"/>
            <rect width="100%" height="400" fill="${riskColor}"/>
            <text x="50%" y="220" font-family="Arial" font-size="80" fill="white" text-anchor="middle" font-weight="bold">${t.title.replace(/&/g, '&amp;')}</text>
            
            <circle cx="540" cy="700" r="200" fill="white" />
            <text x="540" y="740" font-family="Arial" font-size="160" fill="${riskColor}" text-anchor="middle" font-weight="bold">${score}</text>
            
            <rect x="100" y="1000" width="880" height="600" rx="30" fill="white" />
            <text x="540" y="1100" font-family="Arial" font-size="60" fill="#2d3436" text-anchor="middle" font-weight="bold">${t.what_to_do}</text>
            <text x="150" y="1250" font-family="Arial" font-size="45" fill="#636e72">${t.action}: ${data.recommendation?.action || 'N/A'}</text>
            <text x="150" y="1350" font-family="Arial" font-size="45" fill="#636e72">${t.treatment}: ${data.recommendation?.treatment?.product_name || 'N/A'}</text>

            <text x="50%" y="1850" font-family="Arial" font-size="35" fill="#b2bec3" text-anchor="middle">PaddyPulse AI Early Warning System</text>
        </svg>`;

        await sharp(Buffer.from(squareSvg)).png().toFile(squarePath);
        await sharp(Buffer.from(portraitSvg)).png().toFile(portraitPath);

        return { cardId, portraitId, squarePath, portraitPath, cardId: cardId }; // Fix mapping
    }

    /**
     * Complete Workflow Service
     * Gathering data -> Generating Files -> Saving to DB
     */
    async generateFullReportStack(farmId, lang = 'en') {
        const db = require('../database');

        return new Promise((resolve, reject) => {
            const sql = `
                SELECT r.*, f.farmer_name as field_name 
                FROM disease_risk_assessments r
                JOIN farms f ON r.user_id = f.user_id -- Adjusted join
                WHERE f.farm_id = ? 
                ORDER BY r.timestamp DESC LIMIT 1
            `;

            db.get(sql, [farmId], async (err, riskData) => {
                if (err || !riskData) return reject(new Error("No data found"));

                const reportData = {
                    field_name: riskData.field_name,
                    risk_score: riskData.overall_risk_score,
                    risk_level: riskData.overall_risk_score > 75 ? 'CRITICAL' : (riskData.overall_risk_score > 50 ? 'HIGH' : 'MEDIUM'),
                    dominant_disease: 'Blast', // In a real system, compute from scan detections
                    recommendation: JSON.parse(riskData.recommendation_json || '{}'),
                    explainability: [
                        "Environmental suitability for fungi is high",
                        "Neighboring zones report infection spread",
                        "Daily risk trend is accelerating"
                    ]
                };

                const pdf = await this.generatePDF(reportData, lang);
                const card = await this.generateWhatsAppCards(reportData, lang);

                const title = lang === 'te' ? 'వరి చిత్ర నివేదిక' : 'Paddy Health Report';
                db.run(
                    "INSERT INTO reports (farm_id, title, file_path, card_path, status) VALUES (?, ?, ?, ?, ?)",
                    [farmId, title, pdf.reportId, card.cardId, 'Ready'],
                    function (err) {
                        if (err) reject(err);
                        else resolve({ reportId: this.lastID, pdf, card });
                    }
                );
            });
        });
    }
}

module.exports = new ReportEngine();

File: server/engine/report_engine.py
import os
import time
import logging
from pathlib import Path
from PIL import Image, ImageDraw, ImageFont
from reportlab.lib.pagesizes import A4
from reportlab.pdfgen import canvas

logger = logging.getLogger("agro-backend")

BASE_DIR = Path(__file__).resolve().parent.parent
UPLOADS_DIR = BASE_DIR / "uploads"
REPORTS_DIR = UPLOADS_DIR / "reports"
REPORTS_DIR.mkdir(parents=True, exist_ok=True)

# ── Source-type label mapping ─────────────────────────────────────────────────
# Maps the sourceType field value → human-readable view label used in the PDF.
SOURCE_TYPE_LABELS = {
    "manual": "Manual View",
    "live":   "Live Camera View",
    "drone":  "Drone View",
}
DEFAULT_SOURCE_TYPE = "manual"


def get_analyzed_view_label(language: str, source_type: str) -> str:
    """Return the correct 'AI Vision: Analyzed X View' string for the PDF."""
    # Validate and normalise source_type; fall back to 'manual' if unrecognised.
    normalised = source_type.strip().lower() if source_type else DEFAULT_SOURCE_TYPE
    view_label = SOURCE_TYPE_LABELS.get(normalised, SOURCE_TYPE_LABELS[DEFAULT_SOURCE_TYPE])

    if language == "te":
        # Telugu: keep the existing pattern but swap the view label
        return f"AI దృష్టి: {view_label}"
    return f"AI Vision: Analyzed {view_label}"


translations = {
    "en": {
        "title": "Paddy Health & Disease Report",
        "date": "Date",
        "field": "Field",
        "risk_score": "Risk Score",
        "dominant_disease": "Dominant Disease",
        "severity": "Severity",
        "status": "Status",
        "where_problem": "Where is the problem? (Infection Map)",
        "what_to_do": "What to do now? (Recommendations)",
        "action": "Action",
        "treatment": "Treatment",
        "dosage": "Dosage",
        "spray_window": "Best Spray Time",
        "why_suggestion": "Why this suggestion?",
        "safety_note": "Safety Note",
        "cost_breakdown": "Financial: Cost Estimation Breakdown",
        "medicine": "Recommended Medicine",
        "total_cost": "Total Application Cost (Est.)",
        "levels": {
            "CRITICAL": "CRITICAL",
            "HIGH": "HIGH",
            "MEDIUM": "MEDIUM",
            "LOW": "LOW"
        }
    },
    "te": {
        "title": "వరి ఆరోగ్యం మరియు తెగులు నివేదిక",
        "date": "తేదీ",
        "field": "పొలం",
        "risk_score": "ప్రమాద స్థాయి",
        "dominant_disease": "ప్రధాన తెగులు",
        "severity": "తీవ్రత",
        "status": "స్థితి",
        "where_problem": "సమస్య ఎక్కడ ఉంది? (మ్యాప్)",
        "what_to_do": "ఇప్పుడు ఏమి చేయాలి? (సూచనలు)",
        "action": "చేయవలసిన పని",
        "treatment": "చికిత్స",
        "dosage": "మోతాదు",
        "spray_window": "పిచికారీ చేయడానికి ఉత్తమ సమయం",
        "why_suggestion": "ఈ సూచన ఎందుకు?",
        "safety_note": "భద్రతా గమనిక",
        "cost_breakdown": "ఖర్చు అంచనా వివరాలు",
        "medicine": "సిఫార్సు చేయబడిన మందు",
        "total_cost": "మొత్తం ఖర్చు (అంచనా)",
        "levels": {
            "CRITICAL": "అత్యంత ప్రమాదకరం",
            "HIGH": "ప్రమాదకరం",
            "MEDIUM": "మధ్యస్థం",
            "LOW": "తక్కువ"
        }
    }
}

class ReportEngine:
    def resolve_image_path(self, raw_path: str) -> str | None:
        """Resolve an image path stored in the DB.
        Handles absolute Windows paths, relative paths, and just filenames.
        Searches both old 'server/uploads' and new 'server_python/uploads' dirs."""
        if not raw_path:
            return None
        
        # Normalize to forward slashes and extract just the filename
        normalized = raw_path.replace('\\', '/')
        filename = normalized.split('/')[-1]
        
        # Candidate paths to check (in priority order)
        candidates = [
            raw_path,                                          # 1. Direct absolute path
            str(UPLOADS_DIR / filename),                       # 2. server/uploads/<filename>
        ]
        
        for candidate in candidates:
            if os.path.exists(candidate):
                logger.info(f"✅ Resolved image: {candidate}")
                return candidate
        
        logger.warning(f"⚠️ Image not found in any location: {filename}")
        return None

    async def generate_pdf(self, data: dict, language: str = 'en', source_type: str = DEFAULT_SOURCE_TYPE) -> dict:
        t = translations.get(language, translations['en'])
        report_id = f"report_{int(time.time() * 1000)}"
        file_path = REPORTS_DIR / f"{report_id}.pdf"
        
        c = canvas.Canvas(str(file_path), pagesize=A4)
        c.setFont("Helvetica-Bold", 24)
        c.drawCentredString(A4[0] / 2.0, 800, t["title"])
        
        c.setFont("Helvetica", 12)
        c.setFillColorRGB(0.4, 0.4, 0.4)
        date_str = time.strftime("%m/%d/%Y")
        field_name = data.get("field_name", "My Farm")
        c.drawCentredString(A4[0] / 2.0, 770, f"{t['date']}: {date_str} | {t['field']}: {field_name}")
        
        c.setStrokeColorRGB(0.8, 0.8, 0.8)
        c.line(50, 750, A4[0] - 50, 750)
        
        # Risk Section
        risk_level = data.get("risk_level", "LOW")
        score = data.get("risk_score", 0)
        
        c.setFont("Helvetica-Bold", 18)
        c.setFillColorRGB(0.1, 0.1, 0.1)
        c.drawString(50, 710, f"{t['risk_score']}: ")
        
        colors = {
            "CRITICAL": (0.84, 0.19, 0.19),
            "HIGH": (0.88, 0.44, 0.33),
            "MEDIUM": (0.99, 0.8, 0.43),
            "LOW": (0, 0.72, 0.58)
        }
        r, g, b = colors.get(risk_level, colors["LOW"])
        c.setFillColorRGB(r, g, b)
        c.drawString(160, 710, f"{score}/100 ({t['levels'].get(risk_level, risk_level)})")
        
        c.setFont("Helvetica", 14)
        c.setFillColorRGB(0.1, 0.1, 0.1)
        c.drawString(50, 680, f"{t['dominant_disease']}: {data.get('dominant_disease', 'None Detected')}")
        
        # Analyzed Image View — label depends on sourceType
        analyzed_view_label = get_analyzed_view_label(language, source_type)
        c.setFont("Helvetica-Bold", 16)
        c.drawString(50, 640, analyzed_view_label)
        y_cursor = 610
        
        # Try annotated image first, then fallback to original image_reference
        raw_annotated = data.get('annotated_image')
        raw_original = data.get('image_reference')
        
        image_path = self.resolve_image_path(raw_annotated) or self.resolve_image_path(raw_original)
        
        if image_path:
            try:
                c.drawImage(image_path, 50, y_cursor - 200, width=500, height=200, preserveAspectRatio=True)
                y_cursor -= 210
                logger.info(f"✅ Drew image in PDF: {image_path}")
            except Exception as e:
                logger.warning(f"Failed to draw image to PDF: {e}")
                c.setFont("Helvetica", 11)
                c.setFillColorRGB(0.6, 0.6, 0.6)
                c.drawString(60, y_cursor - 20, "[Image could not be rendered in PDF]")
                y_cursor -= 30
        else:
            c.setFont("Helvetica", 11)
            c.setFillColorRGB(0.6, 0.6, 0.6)
            c.drawString(60, y_cursor - 20, "[No analyzed image available]")
            y_cursor -= 30

        # Recommendations
        c.setFont("Helvetica-Bold", 16)
        c.drawString(50, y_cursor - 30, t['what_to_do'])
        
        rec = data.get('recommendation', {})
        c.setFont("Helvetica", 12)
        c.drawString(60, y_cursor - 60, f"• {t['action']}: {rec.get('action', 'Monitoring')}")
        
        treatment = rec.get('treatment', {}).get('product_name', 'Tricyclazole 75% WP') if isinstance(rec, dict) and rec.get('treatment') else 'Tricyclazole 75% WP'
        dosage = rec.get('treatment', {}).get('dosage_per_acre', '120g in 200L water') if isinstance(rec, dict) and rec.get('treatment') else '120g in 200L water'
        
        try:
            sw_list = rec.get('best_spray_window', [])
            spray_window = sw_list[0].get('time', '06:00 AM - 09:00 AM') if sw_list else '06:00 AM - 09:00 AM'
        except:
            spray_window = '06:00 AM - 09:00 AM'
            
        c.drawString(60, y_cursor - 80, f"• {t['treatment']}: {treatment}")
        c.drawString(60, y_cursor - 100, f"• {t['dosage']}: {dosage}")
        c.drawString(60, y_cursor - 120, f"• {t['spray_window']}: {spray_window}")
        
        c.save()
        return {"reportId": report_id, "filePath": str(file_path)}

    async def generate_whatsapp_cards(self, data: dict, language: str = 'en') -> dict:
        t = translations.get(language, translations['en'])
        card_id = f"card_{int(time.time() * 1000)}"
        portrait_id = f"portrait_{int(time.time() * 1000)}"
        square_path = REPORTS_DIR / f"{card_id}.png"
        portrait_path = REPORTS_DIR / f"{portrait_id}.png"
        
        score = data.get("risk_score", 0)
        risk_color = (214, 48, 49) if score > 70 else ((225, 112, 85) if score > 40 else (0, 184, 148))
        disease_name = data.get("dominant_disease", "N/A")
        field_name = data.get("field_name", "My Farm")
        
        # Resolve the analyzed farm image (annotated -> original fallback)
        crop_image_path = (
            self.resolve_image_path(data.get('annotated_image')) or
            self.resolve_image_path(data.get('image_reference'))
        )
        
        try:
            try:
                font_title = ImageFont.truetype("arialbd.ttf", 52)
                font_score = ImageFont.truetype("arialbd.ttf", 130)
                font_text  = ImageFont.truetype("arial.ttf", 36)
                font_small = ImageFont.truetype("arial.ttf", 28)
            except IOError:
                font_title = ImageFont.load_default()
                font_score = ImageFont.load_default()
                font_text  = ImageFont.load_default()
                font_small = ImageFont.load_default()

            # ── Square card (1080x1080) ──
            img = Image.new("RGB", (1080, 1080), (245, 246, 250))
            draw = ImageDraw.Draw(img)

            draw.rectangle([0, 0, 1080, 220], fill=risk_color)
            draw.text((540, 115), t["title"].encode("latin1", "ignore").decode("latin1"),
                      font=font_title, fill="white", anchor="mm")
            draw.text((540, 178), field_name.encode("latin1", "ignore").decode("latin1"),
                      font=font_small, fill="white", anchor="mm")

            # Place crop image on right half if available
            if crop_image_path:
                try:
                    ci = Image.open(crop_image_path).convert("RGB")
                    ci.thumbnail((530, 490))
                    img_x = 1080 - ci.width - 10
                    img_y = 235
                    img.paste(ci, (img_x, img_y))
                    draw.text((img_x + ci.width // 2, img_y + ci.height + 14),
                              "AI Analyzed View", font=font_small, fill=(120, 120, 120), anchor="mm")
                except Exception as e:
                    logger.warning(f"Card image embed failed: {e}")

            # Left: Score + disease
            draw.text((35, 300),  str(round(score)), font=font_score, fill=risk_color, anchor="la")
            draw.text((35, 455),  t["risk_score"].encode("latin1","ignore").decode("latin1"),
                      font=font_text, fill=(80, 80, 80), anchor="la")
            draw.text((35, 510),  f"Disease: {disease_name.encode('latin1','ignore').decode('latin1')}",
                      font=font_text, fill=(60, 60, 60), anchor="la")

            draw.rectangle([0, 1020, 1080, 1080], fill=(225, 225, 225))
            draw.text((540, 1050), "Generated by PaddyPulse AI",
                      font=font_small, fill=(150, 150, 150), anchor="mm")

            img.save(square_path)

            # ── Portrait card (1080x1920) ──
            img_p = Image.new("RGB", (1080, 1920), (241, 242, 246))
            draw_p = ImageDraw.Draw(img_p)
            draw_p.rectangle([0, 0, 1080, 400], fill=risk_color)
            draw_p.text((540, 200), t["title"].encode("latin1","ignore").decode("latin1"),
                        font=font_title, fill="white", anchor="mm")
            draw_p.text((540, 315), field_name.encode("latin1","ignore").decode("latin1"),
                        font=font_small, fill="white", anchor="mm")

            draw_p.ellipse([290, 470, 790, 970], fill="white")
            draw_p.text((540, 720), str(round(score)), font=font_score, fill=risk_color, anchor="mm")
            draw_p.text((540, 875), t["risk_score"].encode("latin1","ignore").decode("latin1"),
                        font=font_text, fill=(80, 80, 80), anchor="mm")

            if crop_image_path:
                try:
                    ci_p = Image.open(crop_image_path).convert("RGB")
                    ci_p.thumbnail((960, 500))
                    px = (1080 - ci_p.width) // 2
                    img_p.paste(ci_p, (px, 1030))
                    draw_p.text((540, 1030 + ci_p.height + 22),
                                "AI Analyzed View", font=font_small, fill=(120, 120, 120), anchor="mm")
                except Exception:
                    pass

            draw_p.text((540, 1885), "Generated by PaddyPulse AI",
                        font=font_small, fill=(180, 180, 180), anchor="mm")
            img_p.save(portrait_path)

        except Exception as e:
            logger.error(f"Image Card Error: {e}")
            
        return {"cardId": card_id, "portraitId": portrait_id, "squarePath": str(square_path), "portraitPath": str(portrait_path)}

report_engine = ReportEngine()
